sql/catalog: disallow uncached resolution of draining names

thoszhang opened the pull request:
Previously, when table descriptors were read from the store in the
`UncachedPhysicalAccessor`, we would reject the descriptor if the
requested name didn't match the current name, which happens when the
namespace entry for the old draining name still exists. This change
updates all other descriptor types to have the same behavior.

The old behavior for non-table descriptors is arguably not incorrect,
since both names are valid during the draining names state, but the
behavior should be consistent.

Release justification: Bug fixes to new functionality

Release note (sql change): This is a change from earlier 20.2 beta
releases. After renaming a database, schema, or type, the old names are
immediately inaccessible from other sessions when referred to in schema
change statements as soon as the new name is committed. This matches the
existing behavior for tables.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/54213)
<!-- Reviewable:end -->

Commit:
sql/catalog: disallow uncached resolution of draining names

Previously, when table descriptors were read from the store in the
`UncachedPhysicalAccessor`, we would reject the descriptor if the
requested name didn't match the current name, which happens when the
namespace entry for the old draining name still exists. This change
updates all other descriptor types to have the same behavior.

The old behavior for non-table descriptors is arguably not incorrect,
since both names are valid during the draining names state, but the
behavior should be consistent.

Release justification: Bug fixes to new functionality

Release note (sql change): This is a change from earlier 20.2 beta
releases. After renaming a database, schema, or type, the old names are
immediately inaccessible from other sessions when referred to in schema
change statements as soon as the new name is committed. This matches the
existing behavior for tables.

ajwerner reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 1 of 1 files at r1, 4 of 4 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/54213#-:-MHDCWet8wI5mAkY_Lhd:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained

___
*[pkg/sql/catalog/catalogkv/physical_accessor.go, line 341 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/54213#-MHDBxCt6nqDb5kL9Gg9:-MHDBxCu2kEP3CuIoiym:b-7kl3wt) ([raw file](https://github.com/cockroachdb/cockroach/blob/a2aecfbfde3c2b78623ae1e15b7241f5b5c2d629/pkg/sql/catalog/catalogkv/physical_accessor.go#L341)):*
> ```Go
> 	if desc.GetName() != object {
> 		if flags.Required {
> 			a.tn = tree.MakeTableNameWithSchema(tree.Name(db), tree.Name(scName), tree.Name(object))
> ```

not actionable but makes me wonder to what extent having different concrete types for type names and table names provides us with value.


<!-- Sent from Reviewable.io -->

lucy-zhang commented:
TFTR

bors r+
